#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [990]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [989]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/lydia/OneDrive/Escritorio/Bootcamp/Modelos ML/ML1/src/data/processed/train1.csv",sep=',')

## Data exploration

In [927]:
# train = train.drop(columns=['Unnamed: 0'])

In [993]:
train.shape

(912, 13)

## Data processing

In [930]:
#Convertir marcas en números en base a su value count. Creamos diccionario
marcas = {'Lenovo': 1, 'Dell': 2, 'HP': 3, 'Asus': 4, 'Acer': 5, 'MSI': 6, 'Toshiba': 7}
#Reemplazamos los valores de marcas
train["Company"] = train["Company"].replace(marcas)

In [931]:
#Convertir tipos en números. Creamos diccionario
tipos = {'Notebook': 1, 'Gaming': 2, 'Ultrabook': 3, '2 in 1 Convertible': 4}
#Reemplazamos los valores de tipos
train["TypeName"] = train["TypeName"].replace(tipos)

In [932]:
#Tratamos la columna ScreenResolution para dividirla en 3. Creamos la columna ScreenInfo para obtener la información de la calidad y si es táctil
train["ScreenInfo"] = train["ScreenResolution"].str[:-9]
train['ScreenResolution'].fillna('Unknown', inplace=True)
train["ScreenInfo"].unique()

array(['IPS Panel Full HD ', 'Full HD ', '', 'Full HD / Touchscreen ',
       'IPS Panel Full HD / Touchscreen '], dtype=object)

In [933]:
# Crear la columna "ScreenQuality"
train['ScreenQuality'] = train['ScreenInfo'].str.replace('Touchscreen', '').str.strip()

# Crear la columna "Touchscreen"
train['Touchscreen'] = train['ScreenInfo'].str.contains('Touchscreen').astype(int)

#Cambiar columna "ScreenResolution" a sólo la resolución
train['ScreenResolution'] = train['ScreenResolution'].str[-9:]

In [934]:
#Convertimos ScreenQuality a numérica
calidad = {'':0,'IPS Panel Full HD': 1,'IPS Panel Full HD /': 1, 'Full HD':1, "Full HD /":1}
train["ScreenQuality"] = train["ScreenQuality"].replace(calidad)

In [935]:
#Convertimos ScreenResolution a numérica ordenado por value counts
resolucion = {'1920x1080':1, '1366x768':2}
train["ScreenResolution"] = train["ScreenResolution"].replace(resolucion)

In [936]:
#Quitar los GHz de la CPU
train['Cpu'] = train['Cpu'].str[:-7]
#Separar cada palabra en una columna
CPU =train["Cpu"].str.split(expand = True)

In [937]:
#Todos son Intel Core
# Crear la columna "Cpu_ICore"
train["Cpu_ICore"] = train["Cpu"].str.split().str[2:].apply(lambda x: ' '.join(filter(None, x)))

In [938]:
#Creamos un diccionario con los modelos de CPU y los pasamos a número
intel_core = {'i5 7200U': 1,
               'i7 7700HQ': 2,
               'i3 600': 3,
               'i5 6200U': 4,
               'i5 8250U': 5,
               'i7 7500U': 6,
               'i3 7100U': 7,
               'i7 8550U': 8,
               'i7 6500U': 9,
               'i7 6700HQ': 10,}
train["Cpu"] = train["Cpu_ICore"].replace(intel_core)

In [939]:
#Eliminamos GB de RAM
train["Ram"] = train["Ram"].str.replace("GB", "")
train['Ram'] = train['Ram'].astype(int)

In [940]:
#Nos quedamos solo con la memoria en GB y lo pasamos a int
train["Memory"] = train["Memory"].str.replace('^128GB SSD +.*', '1128')
train["Memory"] = train["Memory"].str.replace('^256GB SSD +.*', '2256')
train["Memory"] = train["Memory"].str.split("G.*").str[0]
train["Memory"] = train["Memory"].str.replace('^1TB.*', '1000')
train["Memory"] = train["Memory"].astype(int)

C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  
C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """


In [941]:
#Convertimos los valores de las tarjetas gráficas a números

graficas = {'Nvidia GeForce GTX 1050 Ti': 1,
        'Intel HD Graphics 620': 2,
        'Intel HD Graphics 520': 3,
        'Intel UHD Graphics 620': 4,
        'Nvidia GeForce GTX 1060': 5,
        'AMD Radeon 530': 6,
        'Nvidia GeForce 940MX': 7,
        'AMD Radeon R5 M430': 8,
        'Nvidia GeForce GTX 1050': 9,
        'Nvidia GeForce GTX 1070': 10}
train["Gpu"] = train["Gpu"].replace(graficas)

In [942]:
#En la columna "OpSys", convertimos los sistemas en números. Creamos diccionario
sistema = {'No OS': 0, 'Windows 10': 1, 'Linux': 2,'Windows 7': 3}
train["OpSys"] = train["OpSys"].replace(sistema)

In [943]:
#En la columna "Weight", eliminar kg y convertirlo a float.
train["Weight"] = train["Weight"].str.replace("kg", "")
train["Weight"] = train["Weight"].astype(float)

In [944]:
#Eliminamos las columnas "ScreenInfo" y "Cpu_ICore"
train = train.drop(labels=["ScreenInfo", "Cpu_ICore" ], axis=1)
train

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
0         1263        4         2    15.6                 1    2   16    2256   
1          272        1         1    15.6                 1    1    4    1000   
2           26        2         1    15.6                 2    3    4    1000   
3          786        5         1    14.0                 1    7    4     128   
4          404        1         1    14.0                 1    1    8     256   
..         ...      ...       ...     ...               ...  ...  ...     ...   
178        201        3         1    13.3                 1    5    8     512   
179       1184        1         1    15.6                 1    1    8     256   
180        410        5         1    15.6                 1    2    8     256   
181        951        6         2    15.6                 1    2    8    1128   
182       1257        1         1    13.3                 1    6   16     512   

     Gpu  OpSys  Weight  Price_euros  ScreenQuality  Touchscreen  
0      1      1    2.50      1600.00              1            0  
1      2      0    2.20       467.00              1            0  
2      3      1    2.30       418.64              0            0  
3      2      1    1.80       636.00              1            0  
4      2      1    1.90       938.00              1            0  
..   ...    ...     ...          ...            ...          ...  
178    4      1    1.49       959.00              1            0  
179    2      1    1.95      1390.00              1            0  
180    9      2    2.50       879.00              1            0  
181    9      1    2.20      1199.00              1            0  
182    2      1    1.40      1499.00              1            0  

[183 rows x 14 columns]

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [945]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(183, 13)

In [946]:
y.shape

(183,)

### 2. Dividir X_train, X_test, y_train, y_test

In [947]:
from sklearn.model_selection import train_test_split

In [948]:
X_train, X_validation, y_train, y_validation= train_test_split(X, y, test_size = 0.40, random_state = 42)

In [949]:
X_train

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
27          55        5         1    15.6                 2    7    4    1000   
93        1299        2         1    15.6                 2    6    8    1000   
4          404        1         1    14.0                 1    1    8     256   
32         747        5         1    15.6                 2    3    4     500   
100        806        2         1    15.6                 2    1    8     128   
..         ...      ...       ...     ...               ...  ...  ...     ...   
106        135        5         3    14.0                 1    8    8     256   
14         916        1         1    15.6                 1    1    8     512   
92         132        3         1    17.3                 1    5    8     256   
179       1184        1         1    15.6                 1    1    8     256   
102        327        4         1    14.0                 1    6    8     256   

     Gpu  OpSys  Weight  ScreenQuality  Touchscreen  
27     2      1    2.40              0            0  
93     8      2    2.30              0            0  
4      2      1    1.90              1            0  
32     3      1    2.40              0            0  
100    2      1    1.95              0            0  
..   ...    ...     ...            ...          ...  
106    4      1    1.60              1            0  
14     2      1    1.95              1            0  
92     4      1    2.50              1            0  
179    2      1    1.95              1            0  
102    2      1    2.00              1            0  

[109 rows x 13 columns]

### 3. Asignar el modelo (vacío) a una variable


In [950]:
from sklearn.linear_model import LinearRegression

laptop_model = LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [951]:
laptop_model.fit(X_train, y_train)
#OJO!! si ejecutamos esta celda más de una vez, hacemos overfitting

LinearRegression()

In [952]:
laptop_model.coef_

array([ 2.37783778e-01,  2.10927444e+01,  1.64076886e+02, -9.55215735e+00,
       -1.29963240e+02, -2.08547673e+01,  7.92115350e+01, -6.13007759e-02,
        1.01275928e+01,  2.93837773e+01, -1.05735024e+02,  1.29963240e+02,
       -4.60159491e+02])

In [953]:
laptop_model.intercept_

432.0374833066618

In [954]:
predictions = laptop_model.predict(X_validation)
predictions

array([1892.19993989,  466.74418782,  771.90102234,  885.43046747,
        504.77956167, 1356.84770046,  737.39558122,  678.42689827,
        933.337043  ,  694.7313815 ,  320.31623146,  863.98667274,
        359.48515979, 1741.20664688,  658.58688178, 1710.11488342,
        419.59738213, 1784.27937898,  967.78525807, 1098.28807976,
       1033.28827538,  835.56410258, 1250.04959711,  508.66433794,
        914.45983138,  696.82548779,  736.44575563, 1172.41540214,
       1179.1503159 , 1197.3688756 ,  657.43477693,  567.1596972 ,
        785.3743245 ,  530.05162792,  766.52707447,  508.99863408,
        707.66501162,  932.16984272, 1017.24887263,  362.83279958,
       1073.12579729, 1251.33607478,  647.65600951,  878.77252168,
        403.8039617 ,  761.046689  ,  986.25168286,  641.12352821,
        871.50612414, 1878.38673476, 1038.76963375, 1258.41357796,
        245.09916566, 1143.43686776,  490.11184471,  591.9177595 ,
        542.62880582, 1827.64039552, 1446.73022859, 1804.54379

In [955]:
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error ,mean_squared_error, r2_score
import numpy as np

print("MAE: ", mean_absolute_error(y_validation, predictions))
print("MAPE: ", mean_absolute_percentage_error(y_validation, predictions))
print("MSE: ", mean_squared_error(y_validation, predictions))
print("RMSE: ", np.sqrt(mean_squared_error(y_validation, predictions)))
print("R2: ", laptop_model.score(X_validation, y_validation))

MAE:  205.3357533455489
MAPE:  0.25224033925033434
MSE:  65719.97824930224
RMSE:  256.35908068430547
R2:  0.6430648002238015


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [956]:
X_pred = pd.read_csv("data/test.csv")
X_pred.head()

laptop_ID   Company          Product            TypeName  Inches  \
0        750   Toshiba  Tecra X40-D-10G            Notebook    14.0   
1        726  Mediacom    Smartbook 142            Notebook    14.0   
2        633  Mediacom    SmartBook 140            Notebook    14.0   
3        363      Dell    Inspiron 7577              Gaming    15.6   
4        319      Asus    VivoBook Flip  2 in 1 Convertible    11.6   

                            ScreenResolution  \
0  IPS Panel Full HD / Touchscreen 1920x1080   
1                IPS Panel Full HD 1920x1200   
2                          Full HD 1920x1080   
3                          Full HD 1920x1080   
4                       Touchscreen 1366x768   

                                    Cpu  Ram              Memory  \
0            Intel Core i5 7200U 2.5GHz  8GB           256GB SSD   
1              Intel Atom Z8350 1.92GHz  4GB            32GB HDD   
2           Intel Atom x5-Z8350 1.44GHz  2GB  32GB Flash Storage   
3           Intel Core i5 7300HQ 2.5GHz  8GB             1TB HDD   
4  Intel Celeron Dual Core N3350 1.1GHz  2GB  32GB Flash Storage   

                       Gpu       OpSys  Weight  
0    Intel HD Graphics 620  Windows 10  1.25kg  
1    Intel HD Graphics 400  Windows 10   1.4kg  
2        Intel HD Graphics  Windows 10   1.4kg  
3  Nvidia GeForce GTX 1050  Windows 10  2.65kg  
4    Intel HD Graphics 500  Windows 10   1.1kg

In [957]:
X_pred.shape

(391, 12)

### Outliers

In [958]:
#Quitamos todas las líneas de "Company" que tengan menos de 5 repeticiones
value_counts = X_pred['Company'].value_counts()
to_remove = value_counts[value_counts < 5].index
X_pred = X_pred[~X_pred['Company'].isin(to_remove)]

In [959]:
#Quitamos todas las líneas que tengan menos de 25 repeticiones
value_counts = X_pred["TypeName"].value_counts()
to_remove = value_counts[value_counts < 25].index
X_pred = X_pred[~X_pred["TypeName"].isin(to_remove)]

In [960]:
#Quitamos todas las líneas que tengan menos de 20 repeticiones
value_counts = X_pred["ScreenResolution"].value_counts()
to_remove = value_counts[value_counts < 20].index
X_pred = X_pred[~X_pred["ScreenResolution"].isin(to_remove)]

In [961]:
#Quitamos todas las líneas que tengan menos de 14 repeticiones
value_counts = X_pred["Cpu"].value_counts()
to_remove = value_counts[value_counts < 14].index
X_pred = X_pred[~X_pred["Cpu"].isin(to_remove)]

In [962]:
#Quitamos todas las líneas que tengan menos de 20 repeticiones
value_counts = X_pred["Ram"].value_counts()
to_remove = value_counts[value_counts < 20].index
X_pred = X_pred[~X_pred["Ram"].isin(to_remove)]

In [963]:
#Quitamos todas las líneas que tengan menos de 10 repeticiones
value_counts = X_pred["Memory"].value_counts()
to_remove = value_counts[value_counts < 10].index
X_pred = X_pred[~X_pred["Memory"].isin(to_remove)]

In [964]:
#Quitamos todas las líneas que tengan menos de 10 repeticiones
value_counts = X_pred["Gpu"].value_counts()
to_remove = value_counts[value_counts < 10].index
X_pred = X_pred[~X_pred["Gpu"].isin(to_remove)]

In [965]:
#Quitamos todas las líneas que tengan menos de 10 repeticiones
value_counts = X_pred["OpSys"].value_counts()
to_remove = value_counts[value_counts < 10].index
X_pred = X_pred[~X_pred["OpSys"].isin(to_remove)]

In [966]:
#Quitamos todas las líneas que tengan menos de 2 repeticiones
value_counts = X_pred["Weight"].value_counts()
to_remove = value_counts[value_counts < 2].index
X_pred = X_pred[~X_pred["Weight"].isin(to_remove)]

In [967]:
X_pred

laptop_ID  Company                             Product  \
0          750  Toshiba                     Tecra X40-D-10G   
3          363     Dell                       Inspiron 7577   
9         1001   Lenovo                  Legion Y520-15IKBN   
12         541     Asus                           ROG Strix   
13         213   Lenovo                  IdeaPad 320-15IKBN   
..         ...      ...                                 ...   
361       1122   Lenovo                      Yoga 500-15ISK   
362         95     Dell                       Inspiron 7577   
363        408     Asus  FX753VD-GC461T (i7-7700HQ/16GB/1TB   
377       1264     Dell                       Latitude 5480   
379        938   Lenovo                       ThinkPad Yoga   

               TypeName  Inches                           ScreenResolution  \
0              Notebook    14.0  IPS Panel Full HD / Touchscreen 1920x1080   
3                Gaming    15.6                          Full HD 1920x1080   
9                Gaming    15.6                IPS Panel Full HD 1920x1080   
12               Gaming    15.6                IPS Panel Full HD 1920x1080   
13             Notebook    15.6                                   1366x768   
..                  ...     ...                                        ...   
361  2 in 1 Convertible    15.6  IPS Panel Full HD / Touchscreen 1920x1080   
362              Gaming    15.6                          Full HD 1920x1080   
363              Gaming    17.3                          Full HD 1920x1080   
377            Notebook    14.0                                   1366x768   
379  2 in 1 Convertible    13.3  IPS Panel Full HD / Touchscreen 1920x1080   

                             Cpu   Ram                Memory  \
0     Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
3    Intel Core i5 7300HQ 2.5GHz   8GB               1TB HDD   
9    Intel Core i7 7700HQ 2.8GHz   8GB  128GB SSD +  1TB HDD   
12   Intel Core i5 7300HQ 2.5GHz   8GB  128GB SSD +  1TB HDD   
13    Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
..                           ...   ...                   ...   
361   Intel Core i5 6200U 2.3GHz   4GB             256GB SSD   
362  Intel Core i5 7300HQ 2.5GHz   8GB             256GB SSD   
363  Intel Core i7 7700HQ 2.8GHz  16GB  256GB SSD +  1TB HDD   
377   Intel Core i5 7200U 2.5GHz   4GB             500GB HDD   
379   Intel Core i5 7200U 2.5GHz   8GB             512GB SSD   

                         Gpu       OpSys  Weight  
0      Intel HD Graphics 620  Windows 10  1.25kg  
3    Nvidia GeForce GTX 1050  Windows 10  2.65kg  
9    Nvidia GeForce GTX 1060  Windows 10   2.5kg  
12   Nvidia GeForce GTX 1060  Windows 10   2.3kg  
13     Intel HD Graphics 620  Windows 10   2.2kg  
..                       ...         ...     ...  
361    Intel HD Graphics 520  Windows 10   2.1kg  
362  Nvidia GeForce GTX 1060  Windows 10  2.65kg  
363  Nvidia GeForce GTX 1050  Windows 10   2.9kg  
377    Intel HD Graphics 620  Windows 10   1.6kg  
379    Intel HD Graphics 620  Windows 10  1.37kg  

[72 rows x 12 columns]

### Transformaciones

In [968]:
#Se elimina la columna Product
X_pred = X_pred.drop(labels="Product", axis=1)

In [969]:
#Convertir marcas en números en base a su value count. Creamos diccionario
marcas = {'Lenovo': 1, 'Dell': 2, 'HP': 3, 'Asus': 4, 'Acer': 5, 'MSI': 6, 'Toshiba': 7}
#Reemplazamos los valores de marcas
X_pred["Company"] = X_pred["Company"].replace(marcas)

In [970]:
#Convertir tipos en números. Creamos diccionario
tipos = {'Notebook': 1, 'Gaming': 2, 'Ultrabook': 3, '2 in 1 Convertible': 4}
#Reemplazamos los valores de tipos
X_pred["TypeName"] = X_pred["TypeName"].replace(tipos)

In [971]:
#Tratamos la columna ScreenResolution para dividirla en 3. Creamos la columna ScreenInfo para obtener la información de la calidad y si es táctil
X_pred["ScreenInfo"] = X_pred["ScreenResolution"].str[:-9]
X_pred['ScreenResolution'].fillna('Unknown', inplace=True)
X_pred["ScreenInfo"].unique()
# Crear la columna "ScreenQuality"
X_pred['ScreenQuality'] = X_pred['ScreenInfo'].str.replace('Touchscreen', '').str.strip()

# Crear la columna "Touchscreen"
X_pred['Touchscreen'] = X_pred['ScreenInfo'].str.contains('Touchscreen').astype(int)

#Cambiar columna "ScreenResolution" a sólo la resolución
X_pred['ScreenResolution'] = X_pred['ScreenResolution'].str[-9:]
#Convertimos ScreenQuality a numérica
calidad = {'':0,'IPS Panel Full HD': 1,'IPS Panel Full HD /': 1, 'Full HD':1, "Full HD /":1}
X_pred["ScreenQuality"] = X_pred["ScreenQuality"].replace(calidad)
#Convertimos ScreenResolution a numérica ordenado por value counts
resolucion = {'1920x1080':1, '1366x768':2}
X_pred["ScreenResolution"] = X_pred["ScreenResolution"].replace(resolucion)

In [972]:
#Quitar los GHz de la CPU
X_pred['Cpu'] = X_pred['Cpu'].str[:-7]
#Separar cada palabra en una columna
CPU =X_pred["Cpu"].str.split(expand = True)
#Todos son Intel Core
# Crear la columna "Cpu_ICore"
X_pred["Cpu_ICore"] = X_pred["Cpu"].str.split().str[2:].apply(lambda x: ' '.join(filter(None, x)))
#Creamos un diccionario con los modelos de CPU y los pasamos a número
intel_core = {'i5 7200U': 1,
               'i7 7700HQ': 2,
               'i3 600': 3,
               'i5 6200U': 4,
               'i5 8250U': 5,
               'i7 7500U': 6,
               'i3 7100U': 7,
               'i7 8550U': 8,
               'i7 6500U': 9,
               'i7 6700HQ': 10,
               'i5 7300HQ': 11}
X_pred["Cpu"] = X_pred["Cpu_ICore"].replace(intel_core)

In [973]:
#Eliminamos GB de RAM
X_pred["Ram"] = X_pred["Ram"].str.replace("GB", "")
X_pred['Ram'] = X_pred['Ram'].astype(int)

In [974]:
#Nos quedamos solo con la memoria en GB y lo pasamos a int
X_pred["Memory"] = X_pred["Memory"].str.replace('^128GB SSD +.*', '1128')
X_pred["Memory"] = X_pred["Memory"].str.replace('^256GB SSD +.*', '2256')
X_pred["Memory"] = X_pred["Memory"].str.split("G.*").str[0]
X_pred["Memory"] = X_pred["Memory"].str.replace('^1TB.*', '1000')
X_pred["Memory"] = X_pred["Memory"].astype(int)

C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  
C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\lydia\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """


In [975]:
#Convertimos los valores de las tarjetas gráficas a números
graficas = {'Nvidia GeForce GTX 1050 Ti': 1,
        'Intel HD Graphics 620': 2,
        'Intel HD Graphics 520': 3,
        'Intel UHD Graphics 620': 4,
        'Nvidia GeForce GTX 1060': 5,
        'AMD Radeon 530': 6,
        'Nvidia GeForce 940MX': 7,
        'AMD Radeon R5 M430': 8,
        'Nvidia GeForce GTX 1050': 9,
        'Nvidia GeForce GTX 1070': 10}
X_pred["Gpu"] = X_pred["Gpu"].replace(graficas)

In [976]:
#En la columna "OpSys", convertimos los sistemas en números. Creamos diccionario
sistema = {'No OS': 0, 'Windows 10': 1, 'Linux': 2,'Windows 7': 3}
X_pred["OpSys"] = X_pred["OpSys"].replace(sistema)

In [977]:
#En la columna "Weight", eliminar kg y convertirlo a float.
X_pred["Weight"] = X_pred["Weight"].str.replace("kg", "")
X_pred["Weight"] = X_pred["Weight"].astype(float)

In [978]:
X_pred.shape

(72, 15)

In [979]:
#Eliminamos las columnas "ScreenInfo" y "Cpu_ICore"
X_pred = X_pred.drop(labels=["ScreenInfo", "Cpu_ICore" ], axis=1)

In [980]:
X_pred.head()

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
0         750        7         1    14.0                 1    1    8     256   
3         363        2         2    15.6                 1   11    8    1000   
9        1001        1         2    15.6                 1    2    8    1128   
12        541        4         2    15.6                 1   11    8    1128   
13        213        1         1    15.6                 2    1    8     256   

    Gpu  OpSys  Weight  ScreenQuality  Touchscreen  
0     2      1    1.25              1            1  
3     9      1    2.65              1            0  
9     5      1    2.50              1            0  
12    5      1    2.30              1            0  
13    2      1    2.20              0            0

**¿Por qué da error?** 

In [981]:
predictions_submit = laptop_model.predict(X_pred)
predictions_submit

array([ 842.82641789,  923.00196429, 1208.81155958,  996.16335359,
        673.08134933, 1322.91085432,  986.32939491, 1194.15609412,
        762.08738892,  861.89281201,  828.83870569,  836.24439938,
       1271.82504161,  597.18332426,  477.30692975,  983.36933272,
       1536.20997257, 1613.23144385, 1190.8267393 ,  410.0395666 ,
        424.21219305,  568.27133839, 1631.95222609, 1176.02211577,
        938.35681902, 1047.57786606, 1092.56891795, 1051.9864812 ,
       1300.19276544, 1319.17399602,  902.15011773, 1355.16137271,
       1108.44009174,  909.51929594, 1058.79793722, 1172.75661121,
       1044.16981748, 1254.16417407, 1088.60661919, 1054.68765048,
       1143.03363893, 1190.76192601, 1042.87202661, 1075.10716294,
       1328.1002507 ,  461.90908748, 1069.33522058, 1143.32352513,
       1658.4446098 , 1149.82979562, 1070.68712613, 1501.51133738,
        839.02609032, 1242.7261155 , 1691.55768564,  606.8394643 ,
       1854.28891876,  576.47122942, 1453.08479615, 1122.43758

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [982]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [983]:
predictions_submit = laptop_model.predict(X_pred)
predictions_submit

array([ 842.82641789,  923.00196429, 1208.81155958,  996.16335359,
        673.08134933, 1322.91085432,  986.32939491, 1194.15609412,
        762.08738892,  861.89281201,  828.83870569,  836.24439938,
       1271.82504161,  597.18332426,  477.30692975,  983.36933272,
       1536.20997257, 1613.23144385, 1190.8267393 ,  410.0395666 ,
        424.21219305,  568.27133839, 1631.95222609, 1176.02211577,
        938.35681902, 1047.57786606, 1092.56891795, 1051.9864812 ,
       1300.19276544, 1319.17399602,  902.15011773, 1355.16137271,
       1108.44009174,  909.51929594, 1058.79793722, 1172.75661121,
       1044.16981748, 1254.16417407, 1088.60661919, 1054.68765048,
       1143.03363893, 1190.76192601, 1042.87202661, 1075.10716294,
       1328.1002507 ,  461.90908748, 1069.33522058, 1143.32352513,
       1658.4446098 , 1149.82979562, 1070.68712613, 1501.51133738,
        839.02609032, 1242.7261155 , 1691.55768564,  606.8394643 ,
       1854.28891876,  576.47122942, 1453.08479615, 1122.43758

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [984]:
X_pred.head()

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
0         750        7         1    14.0                 1    1    8     256   
3         363        2         2    15.6                 1   11    8    1000   
9        1001        1         2    15.6                 1    2    8    1128   
12        541        4         2    15.6                 1   11    8    1128   
13        213        1         1    15.6                 2    1    8     256   

    Gpu  OpSys  Weight  ScreenQuality  Touchscreen  
0     2      1    1.25              1            1  
3     9      1    2.65              1            0  
9     5      1    2.50              1            0  
12    5      1    2.30              1            0  
13    2      1    2.20              0            0

In [985]:
sample = pd.read_csv("data/sample_submission.csv")

In [986]:
X_pred

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
0          750        7         1    14.0                 1    1    8     256   
3          363        2         2    15.6                 1   11    8    1000   
9         1001        1         2    15.6                 1    2    8    1128   
12         541        4         2    15.6                 1   11    8    1128   
13         213        1         1    15.6                 2    1    8     256   
..         ...      ...       ...     ...               ...  ...  ...     ...   
361       1122        1         4    15.6                 1    4    4     256   
362         95        2         2    15.6                 1   11    8     256   
363        408        4         2    17.3                 1    2   16    2256   
377       1264        2         1    14.0                 2    1    4     500   
379        938        1         4    13.3                 1    1    8     512   

     Gpu  OpSys  Weight  ScreenQuality  Touchscreen  
0      2      1    1.25              1            1  
3      9      1    2.65              1            0  
9      5      1    2.50              1            0  
12     5      1    2.30              1            0  
13     2      1    2.20              0            0  
..   ...    ...     ...            ...          ...  
361    3      1    2.10              1            1  
362    5      1    2.65              1            0  
363    9      1    2.90              1            0  
377    2      1    1.60              0            0  
379    2      1    1.37              1            1  

[72 rows x 13 columns]

In [987]:
sample

laptop_ID  Price_euros
0          750          500
1          726          500
2          633          500
3          363          650
4          319          650
..         ...          ...
386        693          650
387       1132          500
388         49          650
389       1243          650
390        924          500

[391 rows x 2 columns]

In [784]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [785]:
# resetear los índices de ambos datasets
X_pred = X_pred.reset_index(drop=True)
sample = sample.reset_index(drop=True)

# seleccionar la columna Laptop_ID de X_pred
laptop_ids = X_pred["laptop_ID"]

# añadir los valores de Laptop_ID a sample
sample["laptop_ID"] = laptop_ids

# guardar el archivo CSV modificado
sample.to_csv("sample_corregido.csv", index=False)

In [786]:
sample_corregido = pd.read_csv("data/sample_corregido.csv")

In [787]:
sample_corregido = sample_corregido.dropna()
sample_corregido["laptop_ID"] = sample_corregido["laptop_ID"].astype(int)
sample_corregido

laptop_ID  Price_euros
0         750          500
1         363          500
2        1001          500
3         541          650
4         213          650
..        ...          ...
67       1122          650
68         95          650
69        408          500
70       1264          650
71        938          650

[72 rows x 2 columns]

In [788]:
X_pred

laptop_ID  Company  TypeName  Inches  ScreenResolution  Cpu  Ram  Memory  \
0         750        7         1    14.0                 1    1    8     256   
1         363        2         2    15.6                 1   11    8    1000   
2        1001        1         2    15.6                 1    2    8    1128   
3         541        4         2    15.6                 1   11    8    1128   
4         213        1         1    15.6                 2    1    8     256   
..        ...      ...       ...     ...               ...  ...  ...     ...   
67       1122        1         4    15.6                 1    4    4     256   
68         95        2         2    15.6                 1   11    8     256   
69        408        4         2    17.3                 1    2   16    2256   
70       1264        2         1    14.0                 2    1    4     500   
71        938        1         4    13.3                 1    1    8     512   

    Gpu  OpSys  Weight  ScreenQuality  Touchscreen  
0     2      1    1.25              1            1  
1     9      1    2.65              1            0  
2     5      1    2.50              1            0  
3     5      1    2.30              1            0  
4     2      1    2.20              0            0  
..  ...    ...     ...            ...          ...  
67    3      1    2.10              1            1  
68    5      1    2.65              1            0  
69    9      1    2.90              1            0  
70    2      1    1.60              0            0  
71    2      1    1.37              1            1  

[72 rows x 13 columns]

In [720]:
sample_corregido.shape

(72, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [789]:
submission = pd.DataFrame({"laptop_ID": sample_corregido["laptop_ID"], "Price_euros": predictions_submit})
# cuando haya eliminado filas, el laptop ID será del csv que tenga los datos limpios ( que coincida laptop id con los datos que nos hemos quedado)

In [790]:
submission.head()

laptop_ID  Price_euros
0        750  1063.811704
1        363   939.903319
2       1001  1130.015741
3        541  1037.483827
4        213   642.142300

In [791]:
submission.shape

(72, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [792]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == X_pred.shape:
        if df_to_submit.columns.all() == X_pred.columns.all():
            if df_to_submit.laptop_ID.all() == X_pred.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission2.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [793]:
chequeator(submission)

Check the number of rows and/or columns and try again

Mensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.
